prototype_source/pt2e_quantizer.rst

```diff
@@ -302,6 +302,114 @@ functions that are used in the example:
+
+2. Using ``SubgraphMatcher``
+--------------------------------
+Because of this, we recommend people to recognize the pattern through ``SubgraphMatcher``, through capturing a ``torch`` IR pattern (with the same program capture used for capturing the floating point model), instead of using the ``aten`` IR pattern directly.
+
+Example::
+
+  def conv_relu_pattern(x, weight, bias):
+      conv = torch.nn.functional.conv2d(x, weight, bias)
+      relu = torch.nn.functional.relu(conv)
+      return relu
+
+  matcher = SubgraphMatcher(conv_relu_pattern)
+  matches = matcher.match(model)
+  for match in matches:
+      # find input and output of the pattern
+      # annotate the nodes
+      inputs, output = _find_input_and_output(match)
+      inputs[0].users[0].meta["quantization_annotation"] = ...
+      inputs[1].users[0].meta["quantization_annotation"] = ...
+      output.meta["quantization_annotation"] = ...
+
+With this, the ``Quantizer`` will still be valid even when the implementation for nn modules and functionals changes, the ``aten`` IR for floating point model will change, but since we capture the pattern again instead of hardcoding the ``aten`` IR for the pattern, we'll get the updated ``aten`` IR as well and will still be able to match the pattern.
+
+One caveat is that if inputs of the pattern has multiple users, we don't have a good way to identify which user node we want to annotate except for checking the aten op target.
+
+Another caveat is that we need to make sure we have an exhaustive list of examples (e.g. 2D, 3D, 4D inputs, real v.s. symbolic inputs, training=True v.s. training=False etc.) for the pattern to make sure cover different possible ``aten`` IR outcomes captured from the ``torch`` IR pattern.
+
+3. Using ``SubgraphMatcherWithNameNodeMap``
+----------------------------------------------
+We also introduced a different SubgraphMatcher util called ``SubgraphMatcherWithNameNodeMap`` to make it easier to query the nodes that people want to annotate.
+
+Example::
+  
+  def conv_relu_pattern(x, weight, bias):
+      conv = torch.nn.functional.conv2d(x, weight, bias)
+      relu = torch.nn.functional.relu(conv)
+      # returns an additional dict that includes a map from name to node that we want to annotate
+      return relu, {"conv": conv, "relu": relu, "x": x, "weight": weight, "bias": bias}
+
+  matcher = SubgraphMatcherWithNameNodeMap(conv_relu_pattern)
```

Comment:
can we actually remove 3? This is just adding confusion because I dont have any realistic scenario where this will be useful. When you want to "annotate" intermediate nodes, meaning "mark" them as "annotated", then that can be done via nodes returned via subgraph matcher anyway. So honestly this is just adding confusion.

Reply:
I think 3 is actually compatible with the uses in 2 as well, and it can give an easier way to access the node, maybe I can remove 2 and talk about how to use 3 to do annotation without referring to internal nodes, it will be something like:

```
input = match["x"]
weight = match["weight"]
input_user = input.users[0]
input_user.meta["..."] = ...
```

Reply:
TBH we are giving 2 options neither of which are solving the problem you describe in the code snippet in a good way. If decomposition is a real problem 3 doesnt help. If decomposition is not a problem, I can buy that 3 is slightly better, in which case I would just keep 3 and remove 2. 3 is strictly a superset of 2 anyway and there is not significant difference in UX between 2 and 3.

Reply:
> 3 is strictly a superset of 2

yeah agree, maybe I should just remove 2?

we do want to protect against decomposition (one op traced into multiple ops) I think, but for some ops functional op is the same as aten already: https://github.com/pytorch/pytorch/blob/main/torch/nn/functional.py#L2011, unless we expect these will change in the future, we don't necessarily need to worry about them being traced into multiple ops actually.

If we can refer to intermediate node, it's slightly easier to do the annotate I think. But to be more robust, I can also just say always use 3 and always annotate input_users, instead of looking at intermediate node. do you think we should do this then?

Reply:
> If we can refer to intermediate node, it's slightly easier to do the annotate I think. But to be more robust, I can also just say always use 3 and always annotate input_users, instead of looking at intermediate node. do you think we should do this then?

That sounds good to me.